Fix label rendering bug in WebGL1 contexts

jjhembd opened the pull request:
<!--
Thanks for the Pull Request!

Please review [Contribution Guide](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md) before opening your first Pull Request.

To ensure your Pull Request is reviewed and accepted quickly, please refer to our [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/main/CONTRIBUTING.md#pull-request-guidelines).

-->

# Description

This PR fixes a texture loading bug in WebGL1 contexts.

#12100 refactored the `Texture` class for better readability and documentation. In the process, some functions made use of the longer signature for the [texImage2D](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/texImage2D) and [texSubImage2D](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/texSubImage2D) methods, for pixel data supplied as _either a `TypedArray` or an `HTMLImageElement`_. However, only WebGL2 allows the longer signature for `HTMLImageElement` (or similar) data sources. This resulted in a `TypeError` when loading some textures in a WebGL1 context.

The errors came from the `Texture.prototype.copyFrom` method. As part of this fix, the loading calls for different `copyFrom` source types were pulled out into helper functions, for better readability and consistency with the `Texture` constructor.

## Issue number and link

Fixes #12249 

## Testing plan

Load this [local Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=bVDLbsIwEPwVy6cgpY5SCogQUCV6ROJQqb344jgLterYqR9QWuXfaxOgqGUP1u7szHhsrpV1aCdgDwbNkYI9WoIVviEvRyyhmB/npVaOCQWG4hR9U4UQDwh8unXrRDApejCWgQ8P1r1CtZV5gZzxkMZdF85uMKOKqo1XPMoQq+sVq0Amg17fJyGgnHACLAn75GTcaiuipjgnXDLjQsfUkGyMbp5gawBscjcZkXz8kI/HkxQNp2Q6uo81SHsbGa+7SktxfAXFxaU7Ebs+dAzcUfUbdIZTXFp3kLA4ezyKptXGIW9kQkjmoGklC9myyvN3cIRbG20itcyupWUtdkjU8xu/jLhk1obNxkv5LL6A4kWZBf4/qdSsFmq73oGR7BBpb/li1YOEkDIL422l01pWzPxx/gE) and verify that the label renders without errors.

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [x] I have added or updated unit tests to ensure consistent code coverage
- [x] I have updated the inline documentation, and included code examples where relevant
- [x] I have performed a self-review of my code

ggetz commented on packages/engine/Source/Renderer/Texture.js:
```diff
@@ -448,6 +505,35 @@ function loadImageSource(texture, source) {
+  gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, texture.preMultiplyAlpha);
+  gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, texture.flipY);
+
+  gl.texSubImage2D(
```
Maybe I'm just confused by the documentation, but [are height and with not required for a WebGL 2 context](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/texSubImage2D)?

jjhembd replied:
The WebGL1 signature is [still supported in WebGL2](https://registry.khronos.org/webgl/specs/latest/2.0/#3.7.6:~:text=interface%20mixin%20WebGL2RenderingContextOverloads). Basically WebGL2 is [almost 100% backwards compatible with WebGL1](https://webgl2fundamentals.org/webgl/lessons/webgl1-to-webgl2.html), so any function that does not explicitly use WebGL2 features (aside from some extensions) will still work.

ggetz reviewed (changes requested):
Thanks for the fix @jjhembd! 

Would it be possible to add a spec to make sure this regression never happens in the future?

jjhembd commented on packages/engine/Source/Renderer/Texture.js:
```diff
@@ -853,94 +919,33 @@ Texture.prototype.copyFrom = function (options) {
       gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, false);
       gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, false);
-      // initialize the entire texture to zero
-      pixels = PixelFormat.createTypedArray(
-        pixelFormat,
-        pixelDatatype,
-        textureWidth,
-        textureHeight,
-      );
```
After those docs burned me the first time around, I'm now relying on the spec rather than MDN.
The WebGL2 spec only mentions _differences_ from WebGL1. The way `null` is handled is mentioned [here](https://registry.khronos.org/webgl/specs/latest/1.0/index.html#TEXIMAGE2D):

> If pixels is null, a buffer of sufficient size initialized to 0 is passed.

jjhembd commented:
Thanks for the review @ggetz !

> Would it be possible to add a spec to make sure this regression never happens in the future?

The existing `Texture` specs caught the bug, when run in WebGL1. But we have been running them in WebGL2 only.

Since the `Texture` class is so widely used, I think the safest route will be to test _everything_ in both WebGL1 and WebGL2. I updated `TextureSpec` to do this, using the `createWebglVersionHelper` method from #11201.

I think I addressed all the other feedback.

ggetz commented:
Thanks for the updates @jjhembd! I like the plan to run with both WebGL 1 and 2.

I'm seeing some hangups and unexpected failures when running the tests locally. Are you seeing anything similar?

jjhembd commented:
> I'm seeing some hangups and unexpected failures when running the tests locally. Are you seeing anything similar?

Yes, sorry, that was an oversight from https://github.com/CesiumGS/cesium/pull/12301/commits/d385f34702edd1a52b128cb84b2f4a1187f72257. Should be fixed now!

ggetz commented:
Awesome, this is now working for me! Thanks @jjhembd!
